### PR TITLE
Vite: Fix HMR of preview.js

### LINKED ISSUE #21426: [Bug]: Changes to `preview.js` required a manual browser reload to get effect
### Describe the bug

Reported by @tmeasday 

Changing `preview.js` triggered HMR but required a reload. If a project has `preview-head.html` and changes are made in that file, HMR behaves correctly.

### To Reproduce

The reporter used this repo: https://github.com/inkline/inkline

1. Run a Storybook 7 project
2. Make changes to preview.js, e.g.:
```js
// ./preview.js
alert('Hello')
```

The alert should be shown as soon as you save, but it only gives effect once you manually reload the browser page.

### System

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by IanVS**
Closes #21426

## What I did

This makes some changes to the way the vite-builder constructs URLs for preview annotations, and how it handles HMR of those annotations.

Previously, we resolved project files into absolute filesystem paths, because relative paths are not resolvable in the virtual file they are imported in: `virtual:/@storybook/builder-vite/vite-app.js`.  

Now, I've made some changes:
	- Relative files starting with `./` will become just `/`.  This works so long as the CWD is the vite project root, which should normally be the case.  We were using CWD in the require.resolve() previously and didn't hear any complaints, so I don't expect any here either.
	- Absolute paths (for example the `.storybook/preview.js`), will be made relative to the vite project root.
	- Relative files starting with `../` will be made absolute as before.  I don't really know what else to do with those, and I _think_ they either won't happen, or this won't cause problems.

That was the first part of this PR.  The second is handling the actual callback to `import.meta.hot.accept`, which as @tmeasday noticed, probably was never getting hit before and was broken.  Vite only gives us the module that actually changed, the rest are `null`, so we can't use that.  I believe it is no problem to reuse the same function that we use on the initial page load to build up the `getProjectAnnotations` function, as Tom suggested.  So that's what I've done here.

## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Comment by IanVS** [comment]
I added the `windows` label, because at least previously that was needed to get unit tests to run on Windows, which I want.  I'm not sure whether it's needed anymore, so if the tests pass, I'll remove it.

**Comment by tmeasday** [comment]
> - Relative files starting with ./ will become just /. This works so long as the CWD is the vite project root, which should normally be the case. We were using CWD in the require.resolve() previously and didn't hear any complaints, so I don't expect any here either.

All the index entry `importPath` (e.g. for CSF files) are relative to the project root also, so I doubt you could change this and not stuff up the `importFn` in the first place.

File: code/lib/builder-vite/src/utils/process-preview-annotation.ts [hunk 1]
```diff
@@ -37,5 +33,21 @@ export function processPreviewAnnotation(path: PreviewAnnotation | undefined) {
     return transformAbsPath(path);
   }
 
+  // resolve absolute paths relative to project root
+  if (isAbsolute(path)) {
+    return slash(`/${relative(projectRoot, path)}`);
+  }
```
Comment (tmeasday) [inline code/lib/builder-vite/src/utils/process-preview-annotation.ts:39]: What about an absolute path outside the `projectRoot`? Won't that lead to different behaviour to the `../` case below?

**Review by tmeasday — CHANGES_REQUESTED** [review]
One edge case we should probably consider

**Review by tmeasday — APPROVED** [review]

File: code/lib/builder-vite/src/utils/process-preview-annotation.test.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      const annotation = '/Users/foo/storybook/.storybook/preview.js';
+      const url = processPreviewAnnotation(annotation, '/Users/foo/storybook/');
+      expect(url).toBe('/.storybook/preview.js');
+    }
+  );
+
+  it.onWindows(
+    'should convert absolute windows filesystem paths into urls relative to project root',
+    () => {
+      const annotation = 'C:/foo/storybook/.storybook/preview.js';
```
Comment (ndelangen) [inline code/lib/builder-vite/src/utils/process-preview-annotation.test.ts:26]: I think this is actually not a valid path on windows? tagging @valentinpalkovic to confirm.

Comment (valentinpalkovic) [reply code/lib/builder-vite/src/utils/process-preview-annotation.test.ts:26]: Windows, indeed, is often not that strict in resolving absolute paths. The separator can be a forward, a backslash or double forward-slashes from my experience.

**Comment by Pettor** [comment]
@IanVS Recently upgraded a test-project to RC0 and I think I am getting an error related to this fix. When upgrading I know get the following:

```
10:37:51 [vite] Internal server error: Failed to resolve import "C:UsersPetter HancockDevelopmentRepos  emplate-web-app-react
eact/preview" from "..\..\..\..\..\virtual:\@storybook\builder-vite\vite-app.js". Does the file exist?
  Plugin: vite:import-analysis
  File: /virtual:/@storybook/builder-vite/vite-app.js:7:47
  5  |
  6  |    const getProjectAnnotations = async () => {
  7  |      const configs = await Promise.all([import('C:\Users\Petter Hancock\Development\Repos\template-web-app-react\node_modules\@storybook\react/preview'),
     |                                                ^
  8  |  import('@storybook/addon-essentials/docs/preview'),
  9  |  import('@storybook/addon-essentials/actions/preview'),
```

You can see the revolve import path looks broken when Vite tries to resolve it.

**Comment by IanVS** [comment]
@Pettor you're right, @valentinpalkovic and I are currently pairing on a fix for this.

**Pull request merged**